Explain why t-first works with labels (#2157)
ocaml/ocaml.org: data/tutorials/language/0it_04_labels.md

Reviewer: This section should be inserted later. Maybe after [Optional Arguments and Partial Application](https://ocaml.org/docs/labels#optional-arguments-and-partial-application)

Author: Please check the changes in the newest commit.

- : int = 42
```

> However, note that passing labelled arguments throught the pipe operator (`|>`) throws a syntax error:

Reviewer: ```suggestion
**Note**: Passing labelled arguments through the pipe operator (`|>`) throws a syntax error:
```

At parameter definition `~first` is the same as `~first:first`. Passing argument `~last` is the same as `~last:last`.

> ### Applying a function with labelled arguments using the pipe operator

Reviewer: ```suggestion
### Passing Labelled Arguments Using the Pipe Operator
```

> However, labelled arguments can't be passed to the functions using the pipe operator (`|>`):

Reviewer: ```suggestion
Labelled arguments can't be applied to functions throught the pipe operator (`|>`):
```

```ocaml
# let square ~a = a*a;;
val square : a:int -> int = <fun>

# 4 |> square;;
Warning 6 [labels-omitted]: label a was omitted in the application of this function.

- : int = 16
# ~a:4 |> square;;
Error: Syntax error

> ```

Reviewer: ````suggestion
```ocaml
# ~default:42 |> Option.value None;;
Error: Syntax error
```

````

You don't need to define a function here. Only the error is really informative.

The warning
```
Warning 6 [labels-omitted]: label a was omitted in the application of this function.
```

Should be addressed in the [Passing Labelled Arguments](https://ocaml.org/docs/labels#passing-labelled-arguments) section as a note.

Author: Fixed this in the newest commit.

To pass an argument to a function with labelled arguments, we should have at least one unlabelled argument. For instance,

```ocaml
# let double_only_unlabelled ~first second ~third = second*2;;
val double_only_unlabelled : first:'a -> int -> third:'b -> int = <fun>

# 2 |> double_only_unlabelled;;
- : first:'a -> third:'b -> int = <fun>
```

> This, in turn, makes it possible to declare a function with an unlabelled argument as the first one and still be sure, that only this unlabelled argument can be passed using the pipe operator.

Reviewer: This is partly redundant with the note at the end of the [Optional Arguments and Partial Application
](https://ocaml.org/docs/labels#optional-arguments-and-partial-application) section.

Author: Fixed this in the newest commit

> We modify our previous example adding a new parameter **step**.

Reviewer: ```suggestion
Declaring a function's unlabelled argument as the first one simplifies reading the function's type and does not prevent passing this argument using the pipe operator.

Let's modify the `range` function previously defined by adding an additional parameter `step`.
```

At parameter definition `~first` is the same as `~first:first`. Passing argument `~last` is the same as `~last:last`.

> 

Reviewer: ```suggestion
```

**Note**: Optional parameters make it difficult for the compiler to know if a function is partially applied or not. This is why at least one positional parameter is required after the optional ones. If present at application, it means the function is fully applied, if missing, it means the function is partially applied.

### Passing Labelled Arguments Using the Pipe Operator

Labelled arguments can't be applied to functions throught the pipe operator (`|>`):

```ocaml
# ~default:42 |> Option.value None;;
Error: Syntax error

> ```

Reviewer: Isn't this redundant with lines 35-40?

Author: Yes it is, thanks. Fixed this.

Declaring a function's unlabelled argument as the first one simplifies reading the function's type and does not prevent passing this argument using the pipe operator.

Let's modify the `range` function previously defined by adding an additional parameter `step`.

```ocaml
# let rec range step ~first ~last = if first > last then [] else first :: range step ~first:(first + step) ~last;;
val range : int -> first:int -> last:int -> int list = <fun>

# 3 |> range ~last:10 ~first:1;;
- : int list = [1; 4; 7; 10]
```

> 

Reviewer: ```suggestion
```
